include/parallel/parallel_implementation.h

```diff
@@ -1343,10 +1343,10 @@ inline void Communicator::nonblocking_send_packed_range (const unsigned int dest
 
   if (range_begin != range_end)
     {
-//      if (buffer == nullptr)
+      if (buffer == nullptr)
         buffer = std::make_shared<std::vector<buffer_t>>();
```

Comment:
`shared_ptr` is the correct thing here.  Ownership is shared - but it's not with this function... it's with the Request and the caller.  The reference is fine like it is - but I would be fine taking it out too if you think that would make it more clear,

`unique_ptr` doesn't work because the caller needs to maintain a handle to this buffer (so it an re-use it later).

Further: `shared_ptr` is correct because it enables the usage of `SharedPool` ( https://github.com/idaholab/moose/blob/devel/framework/include/utils/SharedPool.h ) to keep track of buffers.  You grab one from the pool - pass it into this function... and then it's automagically returned back to the pool once the parallel request is finished.  It's damn slick and it works well.

Reply:
Oh - as far optimization goes... incrementing the reference counter is in the noise compared to making the parallel request...

Reply:
@permcody You're missing the part where the `shared_ptr` is very much copied into the `PostWaitDereferenceSharedPtr` object down below!

That's when the sharing count goes up by one... and that object holds that reference until the request completes so that no matter if the caller holds it or not it will live for the exact right amount of time.

The caller is free to hold it as well (so they can reuse it later)... but in my use-case I don't have to because the `shared_ptr` automatically returns itself to the SharedPool after `PostWaitDereferenceSharedPtr` goes out of scope at the completion of the parallel request.

Your paying way too much attention to _this_ function's signature... look down below at what happens to that `shared_ptr` to see the rest of the story! 😄

Reply:
Basically: the `shared_ptr` is passed by reference here because this function doesn't need to increase the counter... because it's going to get increased in `PostWaitDereferenceSharedPtr` which is the object that truly holds onto it.  I didn't see the need to increase the count twice and decrement it once... when I could just increase it once.

Reply:
BUT: From an outside API perspective if you think it makes it clearer that the result of calling the function is that ownership will be shared by making the shared_ptr parameter not a reference... then I'm fine changing it.  But it's not at all necessary for correctness.

Reply:
Similarly: If you look at the constructor for `PostWaitDereferenceSharedPtr` you'll notice that I pass in the `shared_ptr` by reference: https://github.com/libMesh/libmesh/blob/master/include/parallel/post_wait_dereference_shared_ptr.h#L41

Once again - this is because there's no reason to increment the counter twice and decrement it once... when just incrementing it once is perfectly fine.

Reply:
OK - Now I see - Thanks!